jerry-debugger/src/lib/debugger-client.ts

```diff
@@ -0,0 +1,404 @@
+
+  constructor(options: JerryDebuggerOptions) {
+    this.host = options.host || DEFAULT_DEBUGGER_HOST;
+    this.port = options.port || DEFAULT_DEBUGGER_PORT;
+    this.verbose = options.verbose || false;
+  }
+
+  connect() {
+    if (this.socket) {
+      this.disconnect();
+    }
+    this.socket = new WebSocket(`ws://${this.host}:${this.port}/jerry-debugger`);
+    this.socket.binaryType = 'arraybuffer';
+    this.socket.on('message', this.onMessage.bind(this));
+
+    return new Promise((resolve, reject) => {
+      if (!this.socket) {
+        reject();
+        return;
+      }
+
+      this.socket.on('open', () => {
+        resolve();
+      });
+
+      this.socket.on('error', (err) => {
+        reject(err);
+      });
+    });
+  }
+
+  disconnect() {
+    if (this.socket) {
+      this.socket.terminate();
+      this.socket = undefined;
+    }
+  }
+
+  abortConnection(message: string) {
+    this.disconnect();
+    console.log('Abort connection: ' + message);
+    throw new Error(message);
+  }
+
+  getFormatSize(format: string) {
+    let length = 0;
+    for (let i = 0; i < format.length; i++) {
+      switch (format[i]) {
+        case 'B':
+          length++;
+          break;
+
+        case 'C':
+          length += this.cpointerSize;
+          break;
+
+        case 'I':
+          length += 4;
+          break;
+
+        default:
+          throw new Error('unsupported message format');
+      }
+    }
+    return length;
+  }
+
+  decodeMessage(format: string, message: Uint8Array, offset = 0) {
+    // Format: B=byte I=int32 C=cpointer
+    // Returns an array of decoded numbers
+
+    const result = [];
+    let value;
+
+    if (offset + this.getFormatSize(format) > message.byteLength) {
+      this.abortConnection('received message too short');
+    }
+
+    for (let i = 0; i < format.length; i++) {
+      if (format[i] === 'B') {
+        result.push(message[offset]);
+        offset++;
+        continue;
+      }
+
+      if (format[i] === 'C' && this.cpointerSize === 2) {
+        if (this.littleEndian) {
+          value = message[offset] | (message[offset + 1] << 8);
+        } else {
+          value = (message[offset] << 8) | message[offset + 1];
+        }
+
+        result.push(value);
+        offset += 2;
+        continue;
+      }
+
+      assert(format[i] === 'I' || (format[i] === 'C' && this.cpointerSize === 4));
+
+      if (this.littleEndian) {
+        value = (message[offset] | (message[offset + 1] << 8)
+                 | (message[offset + 2] << 16) | (message[offset + 3] << 24));
+      } else {
+        value = ((message[offset] << 24) | (message[offset + 1] << 16)
+                 | (message[offset + 2] << 8) | message[offset + 3] << 24);
+      }
+
+      result.push(value);
+      offset += 4;
+    }
+
+    return result;
+  }
+
+  releaseFunction(message: Uint8Array) {
+  }
+
+  getBreakpoint(breakpointData: Array<number>) {
+    const func = this.functions[breakpointData[0]];
```

Comment:
I like your idea of having this `functions` map, but why not do it much easier in the process?
I would try to create a map where the key is the first byte of `message` in `onMessage`.
That way you can make the code a bit cleaner and replace this big switch:

```
  switch (message[0]) { 
```

Reply:
I just read your questions in IRC. Sorry for the delay...

For some reason TS doesn't like using `PROTOCOL.SERVER.XYZ` as key in an `interface`. I'm guessing because property values can be dynamically created at runtime (using JS' `get`ter functionality).

However, it does seem to be OK with simple `const` numbers -- I've not tried, but I'm pretty confident it's also OK when these plain `const`s are imported from another file.

So this worked for me:

```
const JERRY_DEBUGGER_BREAKPOINT_HIT = 16; // change to import from ...
const JERRY_DEBUGGER_EXCEPTION_HIT = 23413243;
const JERRY_DEBUGGER_MEMSTATS = 1231;  

// jerry-dbg-parser.ts

type BytecodeOffset = number;

interface JerryMessageMemstats {
    allocatedBytes: number;
    byteCodeBytes: number;
    stringBytes: number;
    objectBytes: number;
    propertyBytes: number;
}

interface JerryMessageBreakpointHit {
    offset: BytecodeOffset;
}

interface JerryMessageExceptionHit {
    offset: BytecodeOffset;
}

interface ParserFunctionMap {
    [JERRY_DEBUGGER_MEMSTATS]: (Uint8Array) => JerryMessageMemstats;
    [JERRY_DEBUGGER_BREAKPOINT_HIT]: (Uint8Array) => JerryMessageBreakpointHit;
    [JERRY_DEBUGGER_EXCEPTION_HIT]: (Uint8Array) => JerryMessageExceptionHit;
}
```

Reply:
I've gone back and forth trying to understand the point of that FunctionMap you've outlined. I guess to replace the switch statement, you intend that you would indirect through the map to get to a function that will handle each message. One issue is what happens if there is a bogus protocol message with an unexpected 'type'; I guess this can be handled by an exception that there is no such key in the map.

But then if each function in the map returns a different datatype like this, how am I supposed to deal with the return value? So currently I'm having these return void and assume that each function will handle its message type and if useful call a delegate function.